### ISSUE TITLE
dropAll does not drop uid nodes AND go client query can not return the mutation change

### ISSUE THREAD

**Issue by sgon00** [dgraph-io/dgraph#1953]
1. `DropAll` operation does not drop UID nodes. If this is expected, how to delete those empty UID nodes? Is that possible to delete them? 
2. `dg.NewTxn().Query` can not return the mutation change even 10s after the mutation transaction. if this is expected, how can I query the change? 

### Environment

**Dgraph:**
```
Dgraph version   : v1.0.1
Commit SHA-1     : ebce3481
Commit timestamp : 2017-12-20 13:59:58 +0530
Branch           : HEAD
```

**Packages:**
```
$ dep status
PROJECT                      CONSTRAINT  VERSION        REVISION  LATEST   PKGS USED
github.com/dgraph-io/dgraph  *           nightly        2c8f184            3   
github.com/gogo/protobuf     *           v0.5           342cbe0   342cbe0  1   
github.com/golang/protobuf   *           branch master  1e59b77   1e59b77  5   
github.com/pkg/errors        *           v0.8.0         645ef00   645ef00  1   
golang.org/x/net             *           branch master  d866cfc   d866cfc  7   
golang.org/x/text            *           branch master  eb22672   eb22672  14  
google.golang.org/genproto   *           branch master  a8101f2   a8101f2  1   
google.golang.org/grpc       ^1.8.2      v1.8.2         e687fa4   e687fa4  21
```

**Server scripts:**
```
dgraph zero --port_offset -2000 -w ${db_path}/zw/ >> ${log_file_zero} 2>&1 &
sleep 10
dgraph server --memory_mb 1024 --zero localhost:5080 -p ${db_path}/p/ -w ${db_path}/w/ >> ${log_file_server} 2>&1 &
```

**OS:** `Manjaro` (based on `Arch Linux`), 4GM RAM laptop

### main.go

```go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/dgraph-io/dgraph/client"
	"github.com/dgraph-io/dgraph/protos/api"
	"google.golang.org/grpc"
)

type person struct {
	UID     string `json:"uid,omitempty"`
	Name    string `json:"name,omitempty"`
	Age     int    `json:"age,omitempty"`
	Married bool   `json:"married,omitempty"`
}

func main() {
	conn, err := grpc.Dial("127.0.0.1:9080", grpc.WithInsecure())
	if err != nil {
		log.Fatal("While trying to dial gRPC")
	}
	defer conn.Close()

	dg := client.NewDgraphClient(api.NewDgraphClient(conn))

	err = dg.Alter(context.Background(), &api.Operation{DropAll: true})
	handleErr(err)

	err = dg.Alter(context.Background(), &api.Operation{
		Schema: `
		name: string @index(exact) .
		age: int .
		married: bool .
		`,
	})
	handleErr(err)

	p1 := person{Name: "Alice", Age: 26, Married: true}
	p2 := person{Name: "Bob", Age: 27, Married: true}
	p3 := person{Name: "Tom", Age: 28, Married: true}
	p := []person{p1, p2, p3}

	pb, err := json.Marshal(p)
	handleErr(err)

	_, err = dg.NewTxn().Mutate(context.Background(), &api.Mutation{CommitNow: true, SetJson: pb})
	handleErr(err)

	q := `{
		all_persons(func: has(age)) {
			uid
			name
			married
		}
	}`
	resp, err := dg.NewTxn().Query(context.Background(), q)
	handleErr(err)
	fmt.Println("\nQuery all persons: " + string(resp.Json))
	time.Sleep(time.Second * 10)
	fmt.Println("\nQuery all persons: " + string(resp.Json) + "\n")
}

func handleErr(err error) {
	if err != nil {
		log.Fatal(err)
	}
}
```

### The Actual Steps

(1) Run a new dgraph server and zero instance from scratch. (That means `${db_path}` is empty before running dgraph server)
(2) 

```shell
$ go run main.go 

Query all persons: {"all_persons":[]}

Query all persons: {"all_persons":[]}
```

```shell
$ curl -s -X POST localhost:8080/query -d $'
{all_persons(func: has(age)) {
uid
name
married
}}' | jq
{
  "data": {
    "all_persons": [
      {
        "uid": "0x1",
        "name": "Alice",
        "married": true
      },
      {
        "uid": "0x2",
        "name": "Bob",
        "married": true
      },
      {
        "uid": "0x3",
        "name": "Tom",
        "married": true
      }
    ]
  }
}
```

(3)

```shell
$ go run main.go 

Query all persons: {"all_persons":[{"uid":"0x1"},{"uid":"0x2"},{"uid":"0x3"}]}

Query all persons: {"all_persons":[{"uid":"0x1"},{"uid":"0x2"},{"uid":"0x3"}]}
```

```shell
$ curl -s -X POST localhost:8080/query -d $'
{all_persons(func: has(age)) {
uid
name
married
}}' | jq
{
  "data": {
    "all_persons": [
      { 
        "uid": "0x1" 
      },
      { 
        "uid": "0x2" 
      },
      { 
        "uid": "0x3" 
      },
      {
        "uid": "0x4",
        "name": "Alice",
        "married": true
      },
      {
        "uid": "0x5",
        "name": "Bob",
        "married": true
      },
      {
        "uid": "0x6",
        "name": "Tom",
        "married": true
      }
    ]
  }
}
```

(4)

```
$ go run main.go 

Query all persons: {"all_persons":[{"uid":"0x1"},{"uid":"0x2"},{"uid":"0x3"},{"uid":"0x4"},{"uid":"0x5"},{"uid":"0x6"}]}

Query all persons: {"all_persons":[{"uid":"0x1"},{"uid":"0x2"},{"uid":"0x3"},{"uid":"0x4"},{"uid":"0x5"},{"uid":"0x6"}]}
```

```shell
$ curl -s -X POST localhost:8080/query -d $'
{all_persons(func: has(age)) {
uid
name
married
}}' | jq
{
  "data": {
    "all_persons": [
      { 
        "uid": "0x1" 
      },
      { 
        "uid": "0x2" 
      },
      { 
        "uid": "0x3" 
      },
      { 
        "uid": "0x4" 
      },
      { 
        "uid": "0x5" 
      },
      { 
        "uid": "0x6" 
      },
      {
        "uid": "0x7",
        "name": "Alice",
        "married": true
      },
      {
        "uid": "0x8",
        "name": "Bob",
        "married": true
      },
      {
        "uid": "0x9",
        "name": "Tom",
        "married": true
      }
    ]
  }
}
```

**Comment by sgon00** [comment]
PS:
I forgot to mention that those empty nodes should not be returned when I query `func: has(age)` because they don't have `age` predicate.

**Comment by janardhan1993** [comment]
Fixed in master.

**Comment by sgon00** [comment]
@janardhan1993 , I have one more question regarding query can not return mutation change. I tried the following codes in v1.0.1:

```go
_, err := dg.NewTxn().Mutate(ctx, &api.Mutation{CommitNow: true, SetJson: newPersons})
resp, err = dg.NewTxn().Query(ctx, q)
```

I find out even if the query is put after the mutation, the query might complete before the mutation is done.  *(The query result might not have the database change that the above mutation made.)* I am wondering if this is expected behavior or the same bug I found and fixed in this issue. The differences btw this new test and the bug issue is: 

* This bug query can not return the mutation change even 10s after the mutation transaction, 
* But I tried to add sleep 2s btw the mutation and query in this new test code, the query can return the mutation change after the 2s sleep.

(I am not sure if I express myself clearly in English or not) 
If my above words make sense to you, I have three questions:

* Is this behavior expected or a bug?
* If it's a bug, is this the same bug as this issue and fixed already in master?
* If it's not a bug, then how can I run a query to ensure that it's executed after the database is updated?

Thanks a lot.
